packages/frontend/src/model/document.ts

```diff
@@ -19,6 +21,44 @@ export type ModelDocument = Document<"model"> & {
     notebook: Notebook<ModelJudgment>;
 };
 
+/**
+Going from a ModelDocument to a catlog model throws away info. In the other
+direction, if we want to use a catlog model to update an existing ModelDocument,
+then we need to update the cells based on UUID (and possibly add new ones for 
+UUIDs in `updated_model` not found in the current document)
+*/
+export function updateFromCatlogModel(
+    changeDoc: (f: ChangeFn<ModelDocument>) => void,
+    updated_model: DblModel,
+): void {
+    const object_types = new Map<string, string | MorType>();
+    const mor_types = new Map<string, string | ObType>();
+    for (const ob of updated_model.objects()) {
+        object_types.set(ob.content.toString().substring(0, 16), updated_model.obType(ob).content);
```

Comment:
I am alarmed that you're having to call `content.toString()` and then extract a substring here and below.

I assume that this is to work around the sad fact that JavaScript Maps can only have strings as keys. A few comments:

- A heuristic I try to follow is that any algorithm requiring a hash map with non-string keys is too elaborate for JavaScript and should be in the Rust code
- However, I haven't always followed my own advice: if you look in `catlog-wasm/src/theory.rs` you will find helper types `ObTypeIndex` and `MorTypeIndex` that wrap `HashMap<ObType, usize>` and `HashMap<MorType, usize>`. Using these types is another strategy to clean up this code.

Reply:
So the PR is now behaving correct - but this issue still remains. Maybe @jmoggr can help out here? 

The idea is that our livedoc has been converted into a rust model object, a computation happened in rust and returned an updated model. This model should change the object *types* and the morphism *types* but the UUIDs of, e.g., such and such particular arrow in a causal loop diagram, shouldn't have changed. So we can use the fact that those UUIDs haven't changed in order to use the updated *rust* model to modify the present livedoc. This almost works but the UUIDs are off for some reason - things work but if I truncate the last few digits of the UUID.

Reply:
Ok, the types are a bit funky if I don't coerce to string, but it works:

```typescript 
    const object_types = new Map<string | Mor, string | MorType>();
    const mor_types = new Map<
        string | { dom: Mor; cod: Mor; pre: Mor; post: Mor },
        string | ObType
    >();
```